### PR TITLE
Auto generate memory.x

### PR DESCRIPTION
**Pull request by Piroro-hs**
This PR adds `ld` feature, which enables the `memory.x` generation and prevent issues like #215.

### CODE REVIEW

File: build.rs
Comment (Sh3Rm4n) [inline build.rs:13]: Hm. Isn't an enum a better option here? Then the following code should be much more readable.

File: src/lib.rs
Comment (Sh3Rm4n) [inline src/lib.rs:77]: This list gets longer and longer. I think we should reduce this error message to the essential minimum and point to the documentation, what devices / features are supported.

File: .github/workflows/ci.yml
Comment (Sh3Rm4n) [inline .github/workflows/ci.yml:18]: Why are these currently commented out?

File: Cargo.toml
Comment (Sh3Rm4n) [inline Cargo.toml:79]: Out of curiosity, where to have the namings from? Can you point me towards the according documentation?

File: build.rs
Comment (Sh3Rm4n) [inline build.rs:22]: It would be nice to add a comment, where you found these numbers.

File: src/lib.rs
Comment (Sh3Rm4n) [reply src/lib.rs:77]: Note: The `TODO` in line 66 is no longer true. I tripped up on my own TODO note. If you like to, you can remove it :)

File: src/lib.rs
Comment (Sh3Rm4n) [inline src/lib.rs:113]: Because missing `device-selected` does throw a `compile_error!` above already, I wonder if this check is needed at all. 
You could remove it, to save us some indentation :wink:

**Review by Sh3Rm4n — CHANGES_REQUESTED** [review]
I like, that the current `memory.x` file is removed now. It was only added to be able to build examples, but was not intended to be used by the user. Having the `ld` feature for an opt in **correct** `memory.x` files is the correct way to go, IMO :+1: 

I guess this feature is worth a mention in the CHANGELOG.md? :)

File: src/lib.rs
Comment (Rahix) [reply src/lib.rs:113]: I suggest gating everything on `device-selected` anyway.  This way, if no device is selected, the only error message is the one from `compile_error!` and there aren't millions of "item not found" errors alongside it which clutter the compiler output and make it hard to see the original error.

File: src/lib.rs
Comment (Rahix) [reply src/lib.rs:77]: Maybe this message can be compacted without dropping the convenience of listing all supported devices by putting multiple names on a single line?  For example:

```rust
compile_error!(
    "This crate requires you to specify your target chip as a feature.

    Please select one of the following (`x` denotes any character in [a-z]):

        stm32f301x6 stm32f301x8
        stm32f318x8
        stm32f302x6 stm32f302x8 stm32f302xb stm32f302xc stm32f302xd stm32f302xe
        stm32f303x6 stm32f303x8 stm32f303xb stm32f303xc stm32f303xd stm32f303xe
        stm32f328x8
        stm32f358xc
        stm32f398xe
        stm32f373x8 stm32f373xb stm32f373xc
        stm32f378xc
        stm32f334x4 stm32f334x6 stm32f334x8

    Example: The STM32F3Discovery board has a STM32F303VCT6 chip.
    So you want to expand your call to `cargo` with `--features stm32f303xc`.
    For more information, see README -> Selecting the right chip.
    "
);
```

Comment (Sh3Rm4n) [reply src/lib.rs:77]: That would be a good alternative. I'm fine with both.

File: src/lib.rs
Comment (Sh3Rm4n) [reply src/lib.rs:113]: Good point. Than keep it as is :)

File: Cargo.toml
Comment (Piroro-hs) [reply Cargo.toml:79]: Well, 4/6/8/b/c/d/e is just came from part number

File: src/lib.rs
Comment (Piroro-hs) [reply src/lib.rs:113]: I found that we should check `device-selected` in `build.rs`, to avoid `thread 'main' panicked at 'internal error: entered unreachable code', build.rs:39:9` when enable `ld` feature without specifying device (`cargo check --features ld`).

File: Cargo.toml
Comment (Sh3Rm4n) [reply Cargo.toml:79]: I see 😄

File: src/lib.rs
Comment (Sh3Rm4n) [reply src/lib.rs:113]: Either that, or throwing an error, which gives a clear hint to the cause. 

But checking if `device-selected` was set at the before calliong`gen_memory_x` makes sense. Anyway, I would still prefer clearer error-handling instead of `unreachable!` in the `_`-arm in the match cases.

File: build.rs
Comment (Sh3Rm4n) [inline build.rs:4]:
```suggestion
    if cfg!(feature = "ld") and cfg!(feature = "device-selected" {
```

as discussed in https://github.com/stm32-rs/stm32f3xx-hal/pull/216#discussion_r620266515

File: build.rs
Comment (Sh3Rm4n) [inline build.rs:30]: Replace with cleaner error handling, as discussed in https://github.com/stm32-rs/stm32f3xx-hal/pull/216#discussion_r620266515

File: build.rs
Comment (Sh3Rm4n) [inline build.rs:80]: Even though this is clearly unreachable for the current state of the file, I'd prefer an error message + abort / exit(1), which would be much more helpful for the user, if this line is somehow executed in the future.

File: src/lib.rs
Comment (Sh3Rm4n) [inline src/lib.rs:44]:
```suggestion
 So you need to specify `stm32f303xc` in your `Cargo.toml` (note that VC → xc).
```

File: src/lib.rs
Comment (Sh3Rm4n) [inline src/lib.rs:50]: Pretty minor nit (can be ignored) I prefer an empty line after a section header.

**Review by Sh3Rm4n — COMMENTED** [review]
Very much appreciate the documentation updates and moving the device-selected check towards the build.rs file is a good solution as well! :) 

Nice work. Resolving the commented out devices in CI and after that, this looks good to be merged.

File: src/lib.rs
Comment (Piroro-hs) [reply src/lib.rs:44]: Oops

File: .github/workflows/ci.yml
Comment (Piroro-hs) [reply .github/workflows/ci.yml:19]: Devices only differ in memory size are commented out as they only check the literally same code.

Comment (Sh3Rm4n) [reply .github/workflows/ci.yml:19]: Hm. This does make sense. Good catch :+1: Could you at a comment, so this is obvious, when editing the file, so this will not be accidentally deleted? :)

**Comment by Sh3Rm4n** [comment]
I was thinking about maybe adding the `ld` feature to the default features, as well as the `rt` feature. This should be much more user friendly initially and experienced user can turn of default features anyways. 

But I think, this is a discussion for later. Let's first release a version with `memory.x` auto-generation and wait for feedback. Just a note for the future.

**Review by Sh3Rm4n — APPROVED** [review]
LGTM. If you don't say otherwise, I would merge it :)

**Pull request merged**
